scipy/stats/_bootstrap.py

```diff
@@ -0,0 +1,285 @@
+import scipy.stats as stats
+import numpy as np
+from scipy._lib._util import check_random_state
+
+
+def _resample_data(data, n_resamples=None, random_state=None):
+    '''Resample the data'''
+    n = data.shape[-1]
+    if n_resamples:
+        # bootstrap - each row is a random resample of original observations
+        i = random_state.randint(0, n, (n_resamples, n))
+    else:
+        # jackknife - each row leaves out one observation
+        j = np.ones((n, n), dtype=bool)
+        np.fill_diagonal(j, False)
+        i = np.arange(n)
+        i = np.broadcast_to(i, (n, n))
+        i = i[j].reshape((n, n-1))
+    resamples = data[..., i]
+    return resamples
+
+
+def _percentile_of_score(data, score, axis):
+    '''Vectorized, simplified scipy.stats.percentileofscore'''
+    B = data.shape[axis]
+    return (data < score).sum(axis=axis) / B
+
+
+def _percentile_along_axis(theta_hat_b, alpha):
+    '''np.percentile with different percentile for each slice'''
+    shape = theta_hat_b.shape[:-1]
+    alpha = np.broadcast_to(alpha, shape)
+    percentiles = np.zeros_like(alpha)
+    for indices, alpha_i in np.ndenumerate(alpha):
+        theta_hat_b_i = theta_hat_b[indices]
+        percentiles[indices] = np.percentile(theta_hat_b_i, alpha_i)
+    return percentiles
+
+
+def _bca_interval(data, statistic, axis, alpha, theta_hat_b):
+    """Bias-corrected and accelerated interval """
+    # closely follows [1] "BCa Bootstrap CIs"
+    data = data[0]  # only works with 1 sample statistics right now
+
+    # calculate z0_hat
+    theta_hat = statistic(data, axis=axis)[..., None]
+    percentile = _percentile_of_score(theta_hat_b, theta_hat, axis=-1)
+    z0_hat = stats.norm.ppf(percentile)
+
+    # calculate a_hat
+    jackknife_data = _resample_data(data)
+    theta_hat_i = statistic(jackknife_data, axis=-1)
+    theta_hat_dot = theta_hat_i.mean(axis=-1)[..., None]
+    num = ((theta_hat_dot - theta_hat_i)**3).sum(axis=-1)
+    den = 6*((theta_hat_dot - theta_hat_i)**2).sum(axis=-1)**(3/2)
+    a_hat = num / den
+
+    # calculate alpha_1, alpha_2
+    z_alpha = stats.norm.ppf(alpha)
+    z_1alpha = -z_alpha
+    num1 = z0_hat + z_alpha
+    alpha_1 = stats.norm.cdf(z0_hat + num1/(1 - a_hat*(num1)))
+    num2 = z0_hat + z_1alpha
+    alpha_2 = stats.norm.cdf(z0_hat + num2/(1 - a_hat*(num2)))
+    return alpha_1, alpha_2
+
+
+def _bootstrap_ci_iv(data, statistic, axis, confidence_level, n_resamples,
+                     method, random_state):
+    """Input validation for bootstrap_ci"""
+
+    axis_int = int(axis)
+    if axis != axis_int:
+        raise ValueError("`axis` must be an integer.")
+
+    if not data:
```

Comment:
It was useful. Its purpose was to prevent empty iterables from passing through, and non iterables would `TypeError` as not iterable at `for sample in data`. The contents of data are checked in the loop.

I have expanded this to:
```python3
    n_samples = 0
    try:
        n_samples = len(data)
    except TypeError:
        raise ValueError("`data` must be a sequence of samples.")

    if n_samples == 0:
        raise ValueError("`data` must contain at least one sample.")
```
If you think that's worth it, OK.